helix-core/src/normalized_url.rs

```diff
@@ -0,0 +1,175 @@
+        // Since the other fields are derived from this one deterministically, we can simply
+        // compare `.base` to know if two `NormalizedUrl`s are equal.
+        self.base.eq(&other.base)
+    }
+}
+
+impl Eq for NormalizedUrl {}
+
+impl PartialOrd<Self> for NormalizedUrl {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.base.partial_cmp(&other.base)
+    }
+}
+
+impl Ord for NormalizedUrl {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.base.cmp(&other.base)
+    }
+}
+
+impl fmt::Display for NormalizedUrl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.base.fmt(f)
+    }
+}
+
+// Interactions with the original Url type are below
+
+impl From<Url> for NormalizedUrl {
+    fn from(u: Url) -> Self {
+        let path = Path::new(u.path());
+        Self::from_file_path(path).expect("URL failed to convert to absolute path")
+    }
+}
+
+impl PartialEq<Url> for NormalizedUrl {
+    #[cfg(windows)]
+    fn eq(&self, other: &Url) -> bool {
+        self.base.eq(other) || self.opposite_case.as_ref().map_or(false, |oc| oc.eq(other))
+    }
+
+    #[cfg(not(windows))]
+    fn eq(&self, other: &Url) -> bool {
+        self.base.eq(other)
+    }
+}
+
+impl PartialOrd<Url> for NormalizedUrl {
+    fn partial_cmp(&self, other: &Url) -> Option<Ordering> {
+        self.base.partial_cmp(other)
+    }
+}
+
+/// Utility function to get the opposite casing of a Windows path, if any
+#[cfg(windows)]
+fn path_prefix_opposite_casing(default_case_path: &Path) -> Option<PathBuf> {
+    let mut components = default_case_path.components();
+    let first_comp = components.next()?;
+
+    match first_comp {
+        Component::Prefix(prefix)
+        // Anything else than `Disk` variants is not subject to this casing issue
+        if matches!(prefix.kind(), Prefix::VerbatimDisk(_) | Prefix::Disk(_)) => {
+            let cased_prefix = {
+                // Hard to get data on this, but most drive letter should be uppercased by
+                // default so lowercasing first means not having to then uppercase ?
+                let mut prefix_str = prefix.as_os_str().to_ascii_lowercase();
+                if prefix_str == prefix.as_os_str() {
+                    prefix_str.make_ascii_uppercase();
+                }
+                prefix_str
+            };
+
+            // The `push`s should be free thanks to this
+            let mut updated_case_path = PathBuf::with_capacity(default_case_path.as_os_str().len());
+            updated_case_path.push(&cased_prefix);
+            for cp in components {
+                updated_case_path.push(cp.as_os_str());
+            }
+            Some(updated_case_path)
+        },
+        _ => None,
+    }
+}
+
+#[cfg(test)]
+#[cfg(windows)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn different_drives_case_are_equal() {
+        let p1 = Path::new("C:\\one\\path\\Here");
+        let p2 = Path::new("c:\\one\\path\\Here");
```

Comment:
Nice with the tests 🚀

Some more test cases that I can think of off the top of my head:
- Testing `p2` against `p1`
- Testing unix paths
- Testing non-file scheme URLs

Reply:
Done